### ISSUE TITLE
关于 refine retinanet的问题

### ISSUE THREAD

**Issue by PC-god** [Thinklab-SJTU/R3Det_Tensorflow#38]
你好，在你用retinanet做实验时，相关环境可以介绍下么，我用 tf-1.14， cuda-10.0. 
出现了 两个错误：Internal: WhereOp: Could not launch
                              Non-OK-status: CudaLaunchKernel
在构建图的时候正常，执行图的时候run train_op，就会报上面两个错误~

**Comment by yangxue0827** [comment]
我提供了docker镜像，你这个问题试试降低一下tf版本，比如1.12或1.13。

**Comment by PC-god** [comment]
谢谢，我试试！

**Comment by PC-god** [comment]
你好，第一个问题：
        在用 METHOD=R的时候，只能在cpu下跑，如果用gpu版本tf，报WhereOp: Could not launch cub，查询说是tf版本高，cuda要5.1；
在metho=H的时候，可以用10.0的cuda和tf1.13，但是出现：以下nan
per_cost_time:0.653s
cls_loss:1.128
reg_loss:0.556
refine_cls_loss:1.136
refine_reg_loss:nan
total_losses:nan
第二个问题： 在给target_box打标的时候，函数是anchor_target_layer_without_boxweight.py 下面调用的 coordinate_present_convert.py，在坐标转换这个函数里，mode=-1,  是# angle range from [-90, 0) to [-180, 0)，
这里角度范围变化了的目的是什么。 我看里面转换把长边当成w，短边是h，jiao'du。但是最后 为啥  
’‘         coords_new = remain_coords + convert_coords

        coords_new[:, 4] -= 90
’‘
这里将角度 -90 操作没看懂。下面附完整函数``
`def coordinate_present_convert(coords, mode=1):
    """
    :param coords: tensor, shape [-1, 5]
    :param mode: -1 convert coords range to [-90, 90), 1 convert coords range to [-90, 0)
    :return: shape [-1, 5]
    """
    # angle range from [-90, 0) to [-180, 0)
    if mode == -1:
        w, h = coords[:, 2], coords[:, 3]

        remain_mask = np.greater(w, h)
        convert_mask = np.logical_not(remain_mask).astype(np.int32)  # w大的 位置是0
        remain_mask = remain_mask.astype(np.int32)  # w大的 位置是1

        remain_coords = coords * np.reshape(remain_mask, [-1, 1])

        coords[:, [2, 3]] = coords[:, [3, 2]]  # 保证 预测的w是大的，
        coords[:, 4] += 90  #

        convert_coords = coords * np.reshape(convert_mask, [-1, 1])

        coords_new = remain_coords + convert_coords

        coords_new[:, 4] -= 90`

**Comment by PC-god** [comment]
额，第二个问题取消。在cfgs.ANGLE_RANGE == 180的时候，才会用上面的函数。你给的cfg都是90度，不会调这个函数

**Comment by yangxue0827** [comment]
iou_factor = tf.stop_gradient(1-overlaps) / (tf.stop_gradient(regression_loss) + cfgs.EPSILON)

**Comment by PC-god** [comment]
这里，不相当于 loss = -log(iou)  /reg_loss  * reg_loss,  再除以 正样本个数。
没有对reg_loss求范数。 
 tf.stop_gradient（）只是阻止梯度会传，没有像论文中说  reg_loss/ | reg_loss| 来决定梯度方向

**Comment by yangxue0827** [comment]
iou_factor = tf.stop_gradient(1-overlaps) / (tf.stop_gradient(regression_loss) + cfgs.EPSILON)

这句代码只是让iou_factor不传梯度，而不是regression_loss变量不传梯度，如果regression_loss不传梯度，整个网络就不可能训练的起来。regression_loss/ | regression_loss| 可以理解成单位向量，只保留regression_loss来传梯度，loss的幅值由连续的iou来决定

regression_loss = tf.stop_gradient(regression_loss)这样写regression_loss变量才不会反传梯度 @PC-god

**Comment by PC-god** [comment]
呦西，讲究人，谢谢~  最终还是通过reg_loss传梯度，iou负责管大小（也就是你说的幅值）
用起来，iou-smooth-L1下降的能快一点。受教了

**Comment by yangxue0827** [comment]
这么使用原因有两个：
1、解决边界问题
2、近似iou loss（旋转框iou计算麻烦，而且可能是不可导的）

**Comment by PC-god** [comment]
anchors = generate_anchors_pre(6, 6, feat_stride=128, anchor_scales=ANCHOR_SCALES,
                         anchor_ratios=ANCHOR_RATIOS, base_size=512)
在生成anchor时，若'H'，在给generate_anchors.py传参时，base_size固定为4，是不是您笔误？还是我理解有误， base_size是 anchor的基础hw，比如是512，说明ratio=1时，H=W=512

**Comment by PC-god** [comment]
您在训练 R 模式时，环境配置还记得么，H模式都跑通了，R模式失败了~
